python/prairielearn.py

```diff
@@ -93,7 +94,25 @@ def to_json(v):
             M.append(row)
         return {'_type': 'sympy_matrix', '_value': M, '_variables': s, '_shape': [num_rows, num_cols]}
     elif isinstance(v, pandas.DataFrame):
-        return {'_type': 'dataframe', '_value': {'index': list(v.index), 'columns': list(v.columns), 'data': v.values.tolist()}}
+        if new_df_encoding:
+            # The next lines of code are required to address the JSON table-orient
+            # generating numeric keys instead of strings for an index sequence with
+            # only numeric values (c.f. pandas-dev/pandas#46392)
+            df_modified_names = v.copy()
+
+            indexing_dtype = df_modified_names.columns.dtype
+            if indexing_dtype == np.float64 or indexing_dtype == np.int64:
+                df_modified_names.columns = df_modified_names.columns.astype('string')
+
+            # For version 2 storing a data frame, we use the table orientation alongside of
+            # enforcing a date format to allow for numeric values
+            # Details: https://pandas.pydata.org/docs/reference/api/pandas.read_json.html
+            encoded_json_df = df_modified_names.to_json(orient="table", date_format="iso")
+
+            return {'_type': 'dataframe-v2', '_value': encoded_json_df}
```

Comment:
@eliotwrobson this came up from a comment @mwest1066 made in #6501 as the first go around was string encoded with escape sequences instead of native JSON 😢

Reply:
Ah that makes sense. My main question is really if there's a good way of checking this encoding is done properly in the test suite (as opposed to what's there now).

Reply:
@eliotwrobson maybe use `jsonschema` to validate the pandas output? 

I'm not sure we need that to be further scrutinized though. The check now covers the most important part: re-creating the original Pandas data frame.

Reply:
That makes sense. I'll play with the new encoding when I get the chance.